diff --git a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/client/MainnetDataDownloader.js b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/client/MainnetDataDownloader.js
index 2e87b12..056126c 100644
--- a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/client/MainnetDataDownloader.js
+++ b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/client/MainnetDataDownloader.js
@@ -37,11 +37,19 @@ class MainnetDataDownloader {
         // BTW, for most cases, it's the same as Initialization event block number. Let's take this now.
         return this.queryInitializationBlockNumber(poolAddress);
     }
-    async queryInitializationBlockNumber(poolAddress) {
-        let uniswapV3Pool = await this.getCorePoolContarct(poolAddress);
-        let initializeTopic = uniswapV3Pool.filters.Initialize();
-        let initializationEvent = await uniswapV3Pool.queryFilter(initializeTopic);
-        return initializationEvent[0].blockNumber;
+    async queryInitializationBlockNumber(poolAddress, batchSize = 5000) {
+        const pool = await this.getCorePoolContarct(poolAddress);
+        const topic = pool.filters.Initialize();
+        const latest = await this.RPCProvider.getBlockNumber();
+        for (let from = 0; from <= latest; from += batchSize + 1) {
+            console.log(`Searching for Initialize event from block ${from} to ${Math.min(from + batchSize, latest)}`);
+            const to = Math.min(from + batchSize, latest);
+            const logs = await pool.queryFilter(topic, from, to);
+            if (logs.length) {
+                return logs[0].blockNumber;
+            }
+        }
+        throw new Error("Initialize event not found");
     }
     async parseEndBlockTypeWhenInit(toBlock, poolAddress) {
         switch (toBlock) {
@@ -87,7 +95,14 @@ class MainnetDataDownloader {
         if (toBlockAsNumber < deploymentBlockNumber)
             throw new Error(`The pool does not exist at block height: ${toBlockAsNumber}, it was deployed at block height: ${deploymentBlockNumber}`);
         let initializeTopic = uniswapV3Pool.filters.Initialize();
-        let initializationEvent = await uniswapV3Pool.queryFilter(initializeTopic);
+        const initializationWindow = Math.min(batchSize, 9999);
+        let initializationEvent = await uniswapV3Pool.queryFilter(initializeTopic, deploymentBlockNumber, Math.min(deploymentBlockNumber + initializationWindow, toBlockAsNumber));
+        if (!initializationEvent.length) {
+            initializationEvent = await uniswapV3Pool.queryFilter(initializeTopic, deploymentBlockNumber, deploymentBlockNumber);
+        }
+        if (!initializationEvent.length) {
+            throw new Error("Initialize event not found within allowed range.");
+        }
         let initializationSqrtPriceX96 = initializationEvent[0].args.sqrtPriceX96;
         let initializationEventBlockNumber = initializationEvent[0].blockNumber;
         // check db file then
@@ -145,7 +160,14 @@ class MainnetDataDownloader {
             if (0 == initializationEventBlockNumber) {
                 updateInitializationEvent = true;
                 let initializeTopic = uniswapV3Pool.filters.Initialize();
-                let initializationEvent = await uniswapV3Pool.queryFilter(initializeTopic);
+                const initializationWindow = Math.min(batchSize, 9999);
+                let initializationEvent = await uniswapV3Pool.queryFilter(initializeTopic, deploymentBlockNumber, Math.min(deploymentBlockNumber + initializationWindow, toBlockAsNumber));
+                if (!initializationEvent.length) {
+                    initializationEvent = await uniswapV3Pool.queryFilter(initializeTopic, deploymentBlockNumber, deploymentBlockNumber);
+                }
+                if (!initializationEvent.length) {
+                    throw new Error("Initialize event not found within allowed range.");
+                }
                 await eventDB.addInitialSqrtPriceX96(initializationEvent[0].args.sqrtPriceX96.toString());
                 initializationEventBlockNumber = initializationEvent[0].blockNumber;
                 await eventDB.saveInitializationEventBlockNumber(initializationEventBlockNumber);
@@ -212,6 +234,7 @@ class MainnetDataDownloader {
     async downloadEventsFromSubgraph(poolAddress, token0Decimals, token1Decimals, eventDB, fromBlock, toBlock, batchSize) {
         while (fromBlock <= toBlock) {
             let endBlock = fromBlock + batchSize > toBlock ? toBlock : fromBlock + batchSize;
+            console.log(`Downloading SUBGRAPH events from block ${fromBlock} to ${endBlock}`);
             let latestEventBlockNumber = Math.max(await this.saveEventsFromSubgraph(poolAddress, token0Decimals, token1Decimals, eventDB, EventType_1.EventType.MINT, fromBlock, endBlock), await this.saveEventsFromSubgraph(poolAddress, token0Decimals, token1Decimals, eventDB, EventType_1.EventType.BURN, fromBlock, endBlock), await this.saveEventsFromSubgraph(poolAddress, token0Decimals, token1Decimals, eventDB, EventType_1.EventType.SWAP, fromBlock, endBlock));
             await eventDB.saveLatestEventBlockNumber(latestEventBlockNumber);
             fromBlock += batchSize + 1;
@@ -221,6 +244,7 @@ class MainnetDataDownloader {
     async downloadEventsFromRPC(uniswapV3Pool, eventDB, fromBlock, toBlock, batchSize) {
         while (fromBlock <= toBlock) {
             let endBlock = fromBlock + batchSize > toBlock ? toBlock : fromBlock + batchSize;
+            console.log(`Downloading RPC events from block ${fromBlock} to ${endBlock}`);
             let latestEventBlockNumber = Math.max(await this.saveEventsFromRPC(uniswapV3Pool, eventDB, EventType_1.EventType.MINT, fromBlock, endBlock), await this.saveEventsFromRPC(uniswapV3Pool, eventDB, EventType_1.EventType.BURN, fromBlock, endBlock), await this.saveEventsFromRPC(uniswapV3Pool, eventDB, EventType_1.EventType.SWAP, fromBlock, endBlock));
             await eventDB.saveLatestEventBlockNumber(latestEventBlockNumber);
             fromBlock += batchSize + 1;
diff --git a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/core/CorePool.js b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/core/CorePool.js
index e72b9f2..8d4c479 100644
--- a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/core/CorePool.js
+++ b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/core/CorePool.js
@@ -111,10 +111,16 @@ class CorePool {
                 : jsbi_1.default.subtract(TickMath_1.TickMath.MAX_SQRT_RATIO, InternalConstants_1.ONE);
         if (zeroForOne) {
             assert_1.default(jsbi_1.default.greaterThan(sqrtPriceLimitX96, TickMath_1.TickMath.MIN_SQRT_RATIO), "RATIO_MIN");
+            if (!jsbi_1.default.lessThan(sqrtPriceLimitX96, this.sqrtPriceX96)) {
+                sqrtPriceLimitX96 = jsbi_1.default.subtract(this.sqrtPriceX96, InternalConstants_1.ONE);
+            }
             assert_1.default(jsbi_1.default.lessThan(sqrtPriceLimitX96, this.sqrtPriceX96), "RATIO_CURRENT");
         }
         else {
             assert_1.default(jsbi_1.default.lessThan(sqrtPriceLimitX96, TickMath_1.TickMath.MAX_SQRT_RATIO), "RATIO_MAX");
+            if (!jsbi_1.default.greaterThan(sqrtPriceLimitX96, this.sqrtPriceX96)) {
+                sqrtPriceLimitX96 = jsbi_1.default.add(this.sqrtPriceX96, InternalConstants_1.ONE);
+            }
             assert_1.default(jsbi_1.default.greaterThan(sqrtPriceLimitX96, this.sqrtPriceX96), "RATIO_CURRENT");
         }
         const exactInput = jsbi_1.default.greaterThanOrEqual(amountSpecified, InternalConstants_1.ZERO);
diff --git a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/manager/EventDBManager.js b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/manager/EventDBManager.js
index f6e4bfb..417fe8d 100644
--- a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/manager/EventDBManager.js
+++ b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/manager/EventDBManager.js
@@ -201,23 +201,35 @@ class EventDBManager {
         return this.getBuilderContext("liquidity_events", trx)
             .where("type", type)
             .andWhere("date", ">=", startDate)
-            .andWhere("date", "<", endDate);
+            .andWhere("date", "<", endDate)
+            .orderBy("block_number")
+            .orderBy("transaction_index")
+            .orderBy("log_index");
     }
     querySwapEventsByDate(startDate, endDate, trx) {
         return this.getBuilderContext("swap_events", trx)
             .andWhere("date", ">=", startDate)
-            .andWhere("date", "<", endDate);
+            .andWhere("date", "<", endDate)
+            .orderBy("block_number")
+            .orderBy("transaction_index")
+            .orderBy("log_index");
     }
     queryLiquidityEventsByBlockNumber(type, fromBlock, toBlock, trx) {
         return this.getBuilderContext("liquidity_events", trx)
             .where("type", type)
             .andWhere("block_number", ">=", fromBlock)
-            .andWhere("block_number", "<=", toBlock);
+            .andWhere("block_number", "<=", toBlock)
+            .orderBy("block_number")
+            .orderBy("transaction_index")
+            .orderBy("log_index");
     }
     querySwapEventsByBlockNumber(fromBlock, toBlock, trx) {
         return this.getBuilderContext("swap_events", trx)
             .andWhere("block_number", ">=", fromBlock)
-            .andWhere("block_number", "<=", toBlock);
+            .andWhere("block_number", "<=", toBlock)
+            .orderBy("block_number")
+            .orderBy("transaction_index")
+            .orderBy("log_index");
     }
     insertPoolConfig(poolConfig, trx) {
         return this.getBuilderContext("pool_config", trx).insert([
diff --git a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/util/DateUtils.d.ts b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/util/DateUtils.d.ts
index 02b6bb1..067101d 100644
--- a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/util/DateUtils.d.ts
+++ b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/util/DateUtils.d.ts
@@ -2,3 +2,6 @@ export declare function getDate(year: number, month: number, day: number, hour?:
 export declare function getYesterday(date: Date): Date;
 export declare function getTomorrow(date: Date): Date;
 export declare function format(date: Date, fmt: string): string;
+export declare function getNextHour(date: Date): Date;
+export declare function getNextMinute(date: Date): Date;
+export declare function getNext4Hour(date: Date): Date;
diff --git a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/util/DateUtils.js b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/util/DateUtils.js
index a7c3999..e48a1ac 100644
--- a/node_modules/@bella-defintech/uniswap-v3-simulator/dist/util/DateUtils.js
+++ b/node_modules/@bella-defintech/uniswap-v3-simulator/dist/util/DateUtils.js
@@ -15,6 +15,18 @@ exports.getYesterday = getYesterday;
 function getTomorrow(date) {
     return new Date(date.getTime() + 24 * 60 * 60 * 1000);
 }
+function getNextHour(date) {
+    return new Date(date.getTime() + 60 * 60 * 1000);
+}
+exports.getNextHour = getNextHour;
+function getNext4Hour(date) {
+    return new Date(date.getTime() + 4 * 60 * 60 * 1000);
+}
+exports.getNext4Hour = getNext4Hour;
+function getNextMinute(date) {
+    return new Date(date.getTime() + 60 * 1000);
+}
+exports.getNextMinute = getNextMinute;
 exports.getTomorrow = getTomorrow;
 function format(date, fmt) {
     var o = {
